src/daft-dashboard/src/api.rs

```diff
@@ -0,0 +1,268 @@
+    Path(query_id): Path<String>,
+    Json(args): Json<ExecEmitStatsArgs>,
+) -> StatusCode {
+    let mut query_info = state.queries.get_mut(&query_id).unwrap();
+    let QueryState::Executing { exec_info, .. } = &mut query_info.state else {
+        return StatusCode::BAD_REQUEST;
+    };
+
+    for (operator_id, stats) in args.stats {
+        exec_info.operators.get_mut(&operator_id).unwrap().stats = stats;
+    }
+    StatusCode::OK
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct ExecEndArgs {
+    pub exec_end_sec: u64,
+}
+
+async fn exec_end(
+    State(state): State<Arc<DashboardState>>,
+    Path(query_id): Path<String>,
+    Json(args): Json<ExecEndArgs>,
+) -> StatusCode {
+    let mut query_info = state.queries.get_mut(&query_id).unwrap();
+    let QueryState::Executing {
+        exec_info,
+        plan_info,
+    } = &query_info.state
+    else {
+        return StatusCode::BAD_REQUEST;
+    };
+
+    query_info.state = QueryState::Finalizing {
+        plan_info: plan_info.clone(),
+        exec_info: exec_info.clone(),
+        exec_end_sec: args.exec_end_sec,
+    };
+    let summary = query_info.summarize();
+    state.ping_clients(summary);
+    StatusCode::OK
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct FinalizeArgs {
+    pub end_sec: u64,
+    // IPC-serialized RecordBatch
+    pub results: Vec<u8>,
+}
+
+async fn query_end(
+    State(state): State<Arc<DashboardState>>,
+    Path(query_id): Path<String>,
+    Json(args): Json<FinalizeArgs>,
+) -> StatusCode {
+    let mut query_info = state.queries.get_mut(&query_id).unwrap();
+    let QueryState::Finalizing {
+        exec_info,
+        plan_info,
+        exec_end_sec,
+    } = &query_info.state
+    else {
+        return StatusCode::BAD_REQUEST;
+    };
+
+    let Ok(results) = RecordBatch::from_ipc_stream(&args.results) else {
+        return StatusCode::BAD_REQUEST;
+    };
+    query_info.state = QueryState::Finished {
+        plan_info: plan_info.clone(),
+        exec_info: exec_info.clone(),
+        exec_end_sec: *exec_end_sec,
+        end_sec: args.end_sec,
+        results,
+    };
+    let summary = query_info.summarize();
+    state.ping_clients(summary);
+    StatusCode::OK
+}
+
+pub(crate) fn routes() -> Router<Arc<DashboardState>> {
+    Router::new()
+        .merge(client::routes())
+        // Query lifecycle
+        .route("/query/{query_id}/start", post(query_start))
+        .route("/query/{query_id}/plan_start", post(plan_start))
+        .route("/query/{query_id}/plan_end", post(plan_end))
+        .route("/query/{query_id}/exec/start", post(exec_start))
+        .route("/query/{query_id}/exec/{op_id}/start", post(exec_op_start))
+        .route("/query/{query_id}/exec/{op_id}/end", post(exec_op_end))
+        .route("/query/{query_id}/exec/emit_stats", post(exec_emit_stats))
+        .route("/query/{query_id}/exec/end", post(exec_end))
+        .route("/query/{query_id}/end", post(query_end))
```

Comment:
This is interesting, do you have the subscriber posting to all of these, and did you consider a single ws endpoint with a typed message stream?